src/collectives.c

```diff
@@ -675,149 +709,134 @@ shmem_internal_op_to_all_tree(void *target, const void *source, int count, int t
+    const long ps_target_ready = 1, ps_data_ready = 2;
+
+    if (PE_size == 1) {
+        memcpy(target, source, type_size*count);
+        free(current_target);
+        return;
+    }
+
+    while (i != 1) {
+        i >>= 1;
+        pow2_proc <<= 1;
+        log2_proc++;
+    }
 
-         /* perform op */
-         shmem_internal_reduce_local(op, datatype, count,
-                                     target, current_target);
+    /* Currently SHMEM_REDUCE_SYNC_SIZE assumes space for 2^32 PEs; this
+       parameter may be changed if need-be */
+    shmem_internal_assert(log2_proc <= (SHMEM_REDUCE_SYNC_SIZE - 2));
+
+    if (current_target)
+        memcpy(current_target, (void *) source, wrk_size);
+    else
+        RAISE_ERROR_MSG("Failed to allocate current_target (count=%d, type_size=%d, size=%dB)\n",
+                        count, type_size, wrk_size);
+
+    /* Algorithm: reduce N number of PE's into a power of two recursive
+     * doubling algorithm have extra_peers do the operation with one of the
+     * power of two PE's so the information is in the power of two algorithm,
+     * at the end, update extra_peers with answer found by power of two team
+     *
+     * -target is used as "temp" buffer -- current_target tracks latest result
+     * give partner current_result,
+     */
 
-      }
+    /* extra peer exchange: grab information from extra_peer so its part of
+     * pairwise exchange */
+    if (my_id >= pow2_proc) {
+        int peer = (my_id - pow2_proc) * stride + PE_start;
 
-      shmem_internal_quiet();
+        /* Wait for target ready, required when source and target overlap */
+        SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, ps_target_ready);
 
-   /*update extra peer with the final result from the pairwise exchange */
-      if ((PE_size - pow2_proc) > my_id) {
-         peer = (my_id + pow2_proc) * stride + PE_start;
+        shmem_internal_put_nb(target, current_target, wrk_size, peer,
+                              &completion);
+        shmem_internal_put_wait(&completion);
+        shmem_internal_fence();
 
-         shmem_internal_put_nb(target, current_target, wrk_size,
-                               peer, &completion);
-         shmem_internal_put_wait(&completion);
-         shmem_internal_fence();
+        shmem_internal_put_small(pSync_extra_peer, &ps_data_ready, sizeof(long), peer);
+        SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, ps_data_ready);
 
-         buff = neg_one;
-         shmem_internal_put_small(pSync_extra_peer, &buff,
-                                  sizeof(long), peer);
-         shmem_internal_fence();
-      }
+    } else {
+        if (my_id < PE_size - pow2_proc) {
+            int peer = (my_id + pow2_proc) * stride + PE_start;
+            shmem_internal_put_small(pSync_extra_peer, &ps_target_ready, sizeof(long), peer);
+
+            SHMEM_WAIT_UNTIL(pSync_extra_peer, SHMEM_CMP_EQ, ps_data_ready);
```

Comment:
Sorry for the churn in the recdbl_sw code.  I reformatted the code as a separate commit.  If you want to view just the changes, take a look at 320e834.